### ISSUE TITLE
nix build tries to write in nix store

### ISSUE THREAD

**Issue by jooooscha** [nix-community/naersk#236]
First of all, I am not sure if this is to blame on naersk, but as this did not happen with crate2nix I thought I give it a shot.

I am trying to build my [terminal-yt](https://github.com/jooooscha/terminal-yt) with naersk. (On the branch invenstigate-build-error is a minimal flake.nix).
However when running `nix build` I get the following error

<details>
<summary> nix log </summary>
<pre>
"action": "setPhase", "phase": "unpackPhase" }
unpacking sources
unpacking source archive /nix/store/sx08w8rd9r7fg1gn9ryhnxm70yc1qcac-dummy-src
source root is dummy-src
@nix { "action": "setPhase", "phase": "patchPhase" }
patching sources
@nix { "action": "setPhase", "phase": "configurePhase" }
configuring
[naersk] cargo_version (read): 1.60.0
[naersk] cargo_message_format (set): json-diagnostic-rendered-ansi
[naersk] cargo_release: --release
[naersk] cargo_options:
[naersk] cargo_build_options: $cargo_release -j "$NIX_BUILD_CORES" --message-format=$cargo_message_format
[naersk] cargo_test_options: $cargo_release -j "$NIX_BUILD_CORES"
[naersk] RUST_TEST_THREADS: 6
[naersk] cargo_bins_jq_filter: .
[naersk] cargo_build_output_json (created): /build/tmp.tWCrWVuFnh
[naersk] crate_sources: /nix/store/3zdiann5ixfl2jlx166bwkbc76m91h32-crates-io
[naersk] RUSTFLAGS:
[naersk] CARGO_BUILD_RUSTFLAGS:
[naersk] CARGO_BUILD_RUSTFLAGS (updated): --remap-path-prefix /nix/store/3zdiann5ixfl2jlx166bwkbc76m91h32-crates-io=/sources
@nix { "action": "setPhase", "phase": "buildPhase" }
building
cargo build $cargo_release -j "$NIX_BUILD_CORES" --message-format=$cargo_message_format
   Compiling autocfg v1.0.1
   Compiling libc v0.2.80
   Compiling proc-macro2 v1.0.36
   Compiling log v0.4.11
   Compiling cfg-if v0.1.10
   Compiling unicode-xid v0.2.1
   Compiling syn v1.0.84
   Compiling memchr v2.3.4
   Compiling cfg-if v1.0.0
   Compiling pkg-config v0.3.19
   Compiling cc v1.0.67
   Compiling pin-project-lite v0.2.6
   Compiling bytes v1.0.1
   Compiling itoa v0.4.7
   Compiling bitflags v1.2.1
   Compiling serde_derive v1.0.124
   Compiling serde v1.0.124
   Compiling futures-core v0.3.13
   Compiling fnv v1.0.7
   Compiling foreign-types-shared v0.1.1
   Compiling matches v0.1.8
   Compiling lazy_static v1.4.0
  Compiling openssl v0.10.33
   Compiling futures-io v0.3.13
   Compiling native-tls v0.2.7
   Compiling ryu v1.0.5
   Compiling slab v0.4.2
   Compiling futures-task v0.3.13
   Compiling tinyvec_macros v0.1.0
   Compiling httparse v1.3.5
   Compiling futures-sink v0.3.13
   Compiling hashbrown v0.9.1
   Compiling once_cell v1.7.2
   Compiling pin-utils v0.1.0
   Compiling percent-encoding v2.1.0
   Compiling openssl-probe v0.1.2
   Compiling try-lock v0.2.3
   Compiling httpdate v0.3.2
   Compiling tower-service v0.3.1
   Compiling encoding_rs v0.8.28
   Compiling numtoa v0.1.0
   Compiling linked-hash-map v0.5.4
   Compiling serde_json v1.0.64
   Compiling base64 v0.13.0
   Compiling mime v0.3.16
   Compiling tyt v1.2.1 (/build/dummy-src)
   Compiling ipnet v2.3.0
   Compiling cassowary v0.3.0
   Compiling unicode-width v0.1.8
   Compiling unicode-segmentation v1.7.0
   Compiling dtoa v0.4.7
   Compiling alphanumeric-sort v1.4.1
   Compiling tokio v1.3.0
   Compiling indexmap v1.6.2
   Compiling num-traits v0.2.14
   Compiling num-integer v0.1.44
   Compiling futures-channel v0.3.13
   Compiling http v0.2.3
   Compiling foreign-types v0.3.2
   Compiling unicode-bidi v0.3.4
   Compiling tracing-core v0.1.17
   Compiling tinyvec v1.1.1
   Compiling form_urlencoded v1.0.1
   Compiling openssl-sys v0.9.61
   Compiling yaml-rust v0.4.5
   Compiling tracing v0.1.25
   Compiling unicode-normalization v0.1.17
   Compiling http-body v0.4.0
   Compiling want v0.3.0
   Compiling futures-util v0.3.13
   Compiling quote v1.0.9
   Compiling num_cpus v1.13.0
   Compiling mio v0.7.7
   Compiling socket2 v0.3.19
   Compiling dirs-sys-next v0.1.2
   Compiling time v0.1.43
   Compiling termion v1.5.6
   Compiling xcb v0.8.2
   Compiling idna v0.2.2
   Compiling threadpool v1.8.1
   Compiling dirs-next v2.0.0
   Compiling tui v0.12.0
error: failed to run custom build command for `xcb v0.8.2`

Caused by:
  process didn't exit successfully: `/build/dummy-src/target/release/build/xcb-a365697e051dc43c/build-script-build` (exit status: 101)
  --- stderr
  error occured in handler:  ('event:   ', ('xcb', 'Sync', 'AlarmNotify'))
  Traceback (most recent call last):
    File "/nix/store/3zdiann5ixfl2jlx166bwkbc76m91h32-crates-io/xcb-0.8.2/rs_client.py", line 2448, in <module>
      module.generate()
    File "/nix/store/7ryn031i3kjhrhid8hcc3830qkh07h4f-unpack-xcb-0.8.2/xcb-0.8.2/xcbgen/state.py", line 108, in generate
      self.close()
    File "/nix/store/3zdiann5ixfl2jlx166bwkbc76m91h32-crates-io/xcb-0.8.2/rs_client.py", line 296, in rs_close
      _f.writeout(os.path.join(module.rs_srcdir, "ffi", "%s.rs" % _module_name(_ns.ext_name)))
    File "/nix/store/3zdiann5ixfl2jlx166bwkbc76m91h32-crates-io/xcb-0.8.2/rs_client.py", line 114, in writeout
      with open(path, 'w') as f:
  PermissionError: [Errno 13] Permission denied: '/nix/store/3zdiann5ixfl2jlx166bwkbc76m91h32-crates-io/xcb-0.8.2/src/ffi/sync.rs'
  thread 'main' panicked at 'processing of /nix/store/3zdiann5ixfl2jlx166bwkbc76m91h32-crates-io/xcb-0.8.2/xml/sync.xml returned non-zero (1)', /sources/xcb-0.8.2/build.rs:75:17
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
warning: build failed, waiting for other jobs to finish...
error: build failed
[naersk] cargo returned with exit code 101, exiting
</pre>
</details>

Does anyone have an idea where this is coming from and what I could do about it?

Thanks in advance, and thanks for this very nice project (This is the first time I have issues with it :))

**Comment by Patryk27** [comment]
Hi 🙂 

This feels like a bug in naersk - just for kicks, would you mind testing with `singleStep = true;`? (e.g. `naersk.buildPackage { /* some attributes */; singleStep = true; }`).

**Comment by jooooscha** [comment]
Hi, thanks for your response. Unfortunately, this results in the same error

**Comment by Patryk27** [comment]
Okie, then - I'll take a closer look some time this week - thanks for reporting! 🙂

**Comment by Patryk27** [comment]
Hi, I'm back with some answers! 🙂 

tl;dr I'd suggest changing:

```toml
clipboard = "0.5.0"
```
... into:
```toml
copypasta = "0.7"
```
... and then:
```rust
let mut ctx: ClipboardContext = ClipboardContext::new().unwrap();
```
... and voilà, `nix build` works.

A non-tl;dr explanation goes:

`clipboard` (the crate you're using for accessing the clipboard) depends on `xcb`; `xcb` provides a set of low-level functions that allow for applications to talk to the X Server; during compilation, `xcb` uses a [`build.rs` script](https://doc.rust-lang.org/cargo/reference/build-scripts.html) that generates some low-level C bindings (and it's this script failing that causes `nix build` to bail out with an error).

The version of `xcb` used by `clipboard` is pretty old and contains a bug where that build script accidentally stores the auto-generated bindings into the project's directory instead of storing them into a temporary directory created by Cargo; this script happens to work when you run `cargo build` locally (since Cargo doesn't mark dependencies' directories as read-only), but it's technically an invalid thing for a build script to do (and it's already caused some problems in the past, e.g. https://github.com/emilk/egui/issues/146).

The newest version of `xcb` [got that bug fixed](https://github.com/rtbo/rust-xcb/commit/026b58ef0ab928e7f01c356fc9dd4ce29bab3c51#diff-d0d98998092552a1d3259338c2c71e118a5b8343dd4703c0c7f552ada7f9cb42R51), but unfortunately the `clipboard` crate is pretty old and doesn't have an upgrade that'd make it compatible with that newer `xcb` - that's why I'd suggest switching to `copypasta`, which is basically `clipboard` forked to support newer `xcb`; and that's also what makes the compilation work.

Since the issue stems from `xcb` doing a technically invalid thing, I'd consider it a bug there and not in naersk (especially since there's an easy fix in the form of switching to another crate that contains a newer `xcb`) :-)

Once again thanks for reporting and if you encounter any more issues, feel free to re-open this ticket or create a new one.

Cheers;

**Comment by jooooscha** [comment]
Wow, thank you so much for digging so deep into this. And thank you for your detailed and nice answer. Thank you very much
